### ISSUE TITLE
Comment support refactoring approach brainstorm issue

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#638]
- the main problem with comments is that they can be placed much more fine grained than AST nodes. There is no way to deal with these for example: `int a(int b/*x*/)/*x*/{}` because those locations do not have unique nodes. Supplying a comment API based on the AST is always going to be inaccurate, hence all the issues on comment support.
- Lexical preservation code combined with inspecting tokens can be used to manipulate comments, since this algorithm is based on keeping track of tokens, and comments are tokens too. It will be difficult though.
- Javadoc is significant to the AST. Other comments (line and block) not so much. Only keep Javadoc in the AST. Only let Javadoc be a node. It is clear where Javadoc belongs in the AST.
- `CommentsInserter` should not be the default way of dealing with comments. I don't think there should be any default way.
- the pretty printer visitor will lose functionality to print comments. I guess we need to get rid of the visitor and move to a combination of CST+lex pres to get that working again. @ftomassetti ?

Preparation:
- [ ] deprecate `getComment` and friends
- [ ] deprecate `CommentsInserter`
- [ ] copy `CommentsInserter` to `CommentsAttributor`
- [ ] let `CommentsAttributor` return a data structure that maps comments to nodes.
 
4.0.0:
- [ ] delete `CommentsInserter` and the code that is calling it
- [ ] the `comment` field should be removed from the AST
- [ ] the `javadoc` field should be assigned differently (do a post-processing step that sets the Javadoc on any `NodeWithJavadoc` that has a Javadoc comment preceding it directly, as @mernst mentions below)
- [ ] Make `LineComment` and `BlockComment` utility classes for `CommentsAttributor`, thereby taking them out of the AST. I guess `Comment` will go along with it.

**Comment by matozoid** [comment]
Removed "Modifiers should be nodes so that they can have a source position"

**Comment by mernst** [comment]
This approach:

> the javadoc field should be assigned differently (let the parser keep track of the last seen javadoc comment, and assign it every time we create a node that implements NodeWithJavadoc)

is not quite right.  After encountering a Javadoc documentation comment (one that starts with `/**`), if the next non-comment node is a module, package, class, interface, constructor, method, or field declaration, then the Javadoc comment should be attached to it.  However, if the next non-comment node is anything else, then the Javadoc comment should be treated as a normal comment.  In particular, a `/**` comment before an import statement or a local variable definition should be treated as a normal comment.

**Comment by matozoid** [comment]
Thanks, corrected in the top post.

If you or anyone else has any suggestions on how to deal with other comments, that would be useful too.

**Comment by matozoid** [comment]
#1005 and and #1004 will help when manipulating comments without having them associated with a node.

**Comment by ntkoopman** [comment]
The babylon parser has good comment support, as it is designed around code transformations. It would handle `int a(int b/*x*/)/*x*/{}` by adding the first comment as a trailingComment to the Parameter node, and the second one as a leadingComment to BlockStmt. Both trailingComments and leadingComments are arrays.

Wouldn't that work here as well?

**Comment by matozoid** [comment]
@ntkoopman it could, and I even tried it a year ago :-) Two reasons that I gave up on it were...
- there are still situations where you can't say where a comment has to go: `int a/*hello*/(int a, int b)`
- it was seriously going to mess up the grammar: something has to be added for each and every element of every grammar rule. I couldn't see a way to do that reliably.

It's mostly the second point that caused trouble. If someone has a smarter way to insert this info (maybe a postprocessor like the current `CommentsInsert`) - that would be great.

**Comment by ntkoopman** [comment]
Babylon would assign that comment to `SimpleName`.

This project seems to "simplify" the AST in some cases, like `EnumSet<Modifier>` on `MethodDeclaration`. This would make it impossible to assign the comment in `public /*foo*/ static void main` to a node.  In the "official" grammar there would be a `NormalClassDeclaration` with `ClassModifier` nodes. That would also maintain the order.

Changing that would probably be quite a large change though and I don't know if that falls within the project goals.

**Comment by matozoid** [comment]
True, that was a bad example. I've been considering how to do this and I think I'll give it another try.

Second part also true, this is covered in #834. It's not a particularly hard change but will break the API, so we're saving that for the next major version. It was in the way of lexical preservation too.

**Comment by matozoid** [comment]
Something like this:
```
top_comments = empty list

after_list = top comments
before_list = empty list

on token:
	if token is whitespace
		ignore
	else if token is comment
		before_list add token
	else
		add all tokens from before_list to after_list
		clear before_list

on new AST node:
	add all tokens from before_list to node.comments_before
	clear before_list
	after_list = empty list
	node.comments_after = after_list
	(now all comments that aren't "before" the next node will be added to the last node's comments_after list)

on done parsing:
	add all tokens from top_comments to root_node.comments_before
	add all tokens from after_list to root_node.comments_after
	add all tokens from before_list to root_node.comments_after
```

**Comment by Wouter1** [comment]
This issue has been open for 7 years now.... Maybe I can suggest a few things and request the urgence to be raised.
My use case:  I'm transforming the java code. Transformation hints can be supplied by the programmer.
Eg 
```
//transformationhint: 342
/**  @param... @returns blabla */
public String f(...) {
}
```

The transformation hints are in comments because they are not relevant for the java part. Hence no use of annotations or other tricks. Please no discussions about this choice.

From what I see (I tested with 3.24.4) the parser drops the transformationhint and now the transformation fails.

The programmer might be able figure out the problem (hard to start with...) and "fix" it and put the transformationhint below the javadoc, But now the javadoc (also to be transformed) would disappear from the transformation. So a programmer can not work around this issue. Bottom line I have to tell the programmer that he has to drop his javadoc here in favour of the annotation.....

May I propose that if there is a list of comments, just attach them in a chain. Simply add a field to Comment:
```
private Optional<Comment> other
```
and a getter:
```
public Optional<Comment> getOther() {
return other;
}
```

The prettyprinters may need some fixing as well of course (irrelevant for me).

This at least solves the problem of disappearing comments and allows people to at least 'sort it out'. Old code remains working. The default getComment keeps returning the last comment.

>Javadoc is significant to the AST. Other comments (line and block) not so much. Only keep Javadoc in the AST. Only let Javadoc be a node. It is clear where Javadoc belongs in the AST.

I completely disagree. Javadoc is highly formalized, to annotate the meaning of the java code input and output behaviour. block comments and line comments are for other comments. In my case, so transformation hints, should NOT end up in javadoc as it has nothing to do with the behaviour of the java program.

**Comment by jlerbsc** [comment]
Hi, there is nothing planned to handle comments differently. Attached is an article describing how comments are handled in javaparser  https://tomassetti.me/java-comments-parsing/

**Comment by elharo** [comment]
Just pulling a .java file in and writing it pack out again results in moving line ending comments to the next line. E.g. from a diff

```
-        buf.getChar(); // minor + ver
+        // minor + ver
+        buf.getChar();
```

My print config looks like this:

```
        DefaultPrinterConfiguration configuration = new DefaultPrinterConfiguration();
        configuration.addOption(new DefaultConfigurationOption(ConfigOption.END_OF_LINE_CHARACTER, "\n"));
        return compilationUnit.toString(configuration);
```

Possibly there's some config that doesn't do this, but otherwise I might have to explore other options for processing .java files or resort to some very ugly hacks. From the discussion here I thought that much had been fixed, but apparently not. 

I suppose the AST is dropping whitespace and then the printer is reinserting it. I'm no expert parsers by any means. I do wonder if it would be possible or desirable to optionally include whitespace nodes in the AST to enable better round-tripping.

**Comment by jlerbsc** [comment]
Have you used the “lexical preservation printer”?

**Comment by DvonHolten** [comment]
back in the day when i reverse engineered compilers by the Niklaus Wirth book, the tokens in the input stream kept their line/column source position. So you can assemble source lines without having the whitespace.

**Comment by elharo** [comment]
Thanks. I tried the the LexcialPreservingPrinter. I don't have it working quite right yet, but I might be doing something wrong. 

The idea of adding line/column source position to each node sounds interesting, though I'm not sure how it would work with tabs. Line numbers would at least be reproducible though, even if columns could be a little wonky. XML parsers do something similar with a separate Locator interface (either DOM or SAX) that optionally provides this info.
